### PR TITLE
Discussion part 4: comments, nested comments/replies + ability to comment on discussion and comments

### LINKED ISSUE #543: Discussions
This needs:

- [x] A (slight) redesign of the discussion cards in the communities page (Maybe just adding avatars)
- [x] A `Comment` component to display a comment with a reply button (and a text input that gets revealed?)
  - [x] Show replies
  - [x] Reply button reveals form that can be used to reply to that top level comment
- [x] ~~A `Reply` component, a reply to a comment that doesn't have a reply button (there is only 1 level of replies)~~
- [x] A `CommentTree` component that shows
  - [x] a list of comments
  - [x] all of the comments' replies
- [x] A page that lists all the discussions in a community
  - [x] ~~Refactor this later to share more markup with discussion section if it makes sense~~
  - [x] Tests
  - [x] Storybook
- [x] A page that shows the comments/replies for a single discussion
  - [x] Redo to match design
  - [x] Storybook
  - [x] Render single discussion with community markup - see [this PR discussion more details](https://github.com/Couchers-org/couchers/pull/1197#issuecomment-835881976)
- [x] A page for creating a new discussion in a community
  - [x] Functional form
  - [x] Styling
  - [x] Storybook
  - [x] Tests

These components should be in storybook

This will use #536 but can be done with normal text or the existing markdown editing component for now

### PR DESCRIPTION
**Pull request by darrenvong**
Covers more of #543.

Discussion feature as described in PR title.

I'll add screenshots later. Tests left to do but feel free to do an early review if you got time

**Frontend checklist**
- [x] Formatted my code with `yarn format && yarn lint --fix`
- [x] There are no warnings from `yarn lint`
- [x] There are no console warnings when running the app
- [x] Added any new components to storybook
- [x] Added tests where relevant
- [x] All tests pass
- [x] Clicked around my changes running locally and it works
- [x] Checked Desktop, Mobile and Tablet screen sizes

### CODE REVIEW

**Comment by lucaslcode** [comment]
With the URLs, looking at `docs/URLs.py` what you have now is correct. But the community (aka official cluster) id is in the database, so it should be trivial to include it in `GetDiscussion`, so in that case I think we can put the discussion page inside the community. Dunno if you want to do it in this PR or not?

**Comment by darrenvong** [comment]
How would this work with the URL matching though, since the community pages start with `/community` and a single discussion starts with `/discussion`? I'm guessing we'll probably have to duplicate or split the common bits into a shared component so both community and discussion can use it?

Also how would things look roughly layout wise? Instead of the tabs with the different things, we replace it with the discussion in the same place if we are on a single discussion?

**Comment by lucaslcode** [comment]
Yes, I guess the community page would need some refactoring so it can take a child?
The top bit of the community page (everything up to line 190) can be `CommunityBase.tsx`, and it can optionally take a `communityId` prop - if it doesn't have it, it gets the id from the url. And then put `{ children }` just below it.

Then the real community page can have `<CommunityBase>/*<Switch>s here*/</CommunityBase>` and discussions page can be `<CommunityBase communityId={discussion.communityId}>/*discussion page*/</CommunityBase>`

**Comment by lucaslcode** [comment]
Oh and I didn't answer your question - I think what would make the most sense is the tab bar is still there, just none of them are selected

**Comment by darrenvong** [comment]
Hmm I think I'll do the page refactor in another PR, as I imagine this one will get pretty huge otherwise with big diffs as is usually the case when refactoring/shifting things around

**Comment by lucaslcode** [comment]
Didn't notice last time - the cards on the community page are slightly broken on mobile: 
image

**Comment by lucaslcode** [comment]
~~I think the button to add a nested comment should move outside the card and at the bottom of any other nested comments:~~ 
image
~~Otherwise, it's not intuitive what happens when there are already nested comments (like in this image).~~

I see, nested comments are also latest first. Hmm. ~~I wonder if they should be oldest first? That might require backend changes.~~

After thinking about it, how about making the latest comments at the bottom, and displaying a "load earlier comments" link at the top? And indeed moving the button as I have in the image.

There also seems to be a bit too much bottom padding in the cards

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/Comment.tsx]:
```suggestion
        <Avatar user={user} className={classes.avatar} />
```

File: app/frontend/src/features/communities/discussions/CommentForm.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/CommentForm.tsx:62]: I can't remember what you decided last time, but this is suffering from the same problem where on a slow connection, the loading state finishes when the new comment is posted but before it actually appears.

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/Comment.tsx]: How about
```suggestion
    marginInlineStart: theme.spacing(3),
    marginInlineStart: `clamp(${theme.spacing(2)}, 5vw, ${theme.spacing(5)})`,
```
Mobile is most likely not to support clamp, so I made the fallback a bit less

File: app/frontend/src/features/communities/discussions/DiscussionPage.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/DiscussionPage.tsx]: This should be in the relevant components (or more specific) rather than here. It's appropriate for the discussion content, but it's also affecting the cards, where it's too much margin (imo)

File: app/frontend/src/features/communities/discussions/DiscussionPage.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/DiscussionPage.tsx]: remove

**Review by lucaslcode — CHANGES_REQUESTED** [review]
Haven't looked through all the code yet - these are just some design comments

**Comment by darrenvong** [comment]
> I think the button to add a nested comment should move outside the card and at the bottom

Basing off Reddit and Facebook, they all have the reply button right under/near the comment you are replying to, so not too sure about putting it under the last nested comment/reply. That could be unexpected behaviour for users who are used to other sites. What's your main objection on keeping the button inside the comment card in that case?

> how about making the latest comments at the bottom, and displaying a "load earlier comments" link at the top?

I know we did for chat, but since these comments also has a reply button (i.e. something keyboard focusable), I'm not sure it's a good idea to do the same CSS trick of reversing the display order of the comments as keyboard users would probably end up going from bottom to top if they tab between buttons. Getting that to work "out of the box" as we do now might be a bit of a pain.

Reversing the data is a potential option, but feels like extra work that doesn't need to be done by the frontend if the backend gives it in the right order to start with. And the operation is linear time so it's gonna make our page slower as it grows (although unlikely to be an immediate problem for some time) so probably better done in the backend.

**Comment by lucaslcode** [comment]
The difference is, reddit has infinite reply levels, and clicking reply adds another level. Here, it will only go one level deep - I think the difference is important.

Facebook does have the reply button on the comment, but when you click it, it opens the comment box underneath all the other comments. We could do that!

> keyboard users would probably end up going from bottom to top

Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?

File: app/frontend/src/features/communities/discussions/CommentForm.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/CommentForm.tsx]: Would this be better as `alwaysShow` or `toggleable`?

File: app/frontend/src/features/communities/discussions/CommentTree.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/CommentTree.tsx]: Why not `threadId`? There could easily be comments posted in the same second.

**Review by lucaslcode — CHANGES_REQUESTED** [review]

File: app/frontend/src/features/communities/discussions/CommentTree.tsx
Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/CommentTree.tsx]: That's a good point, `threadId` is much better - I remember I was scrambling to find something unique but probably overlooked this with the whole thread/comment confusion at the beginning 😅

File: app/frontend/src/features/communities/discussions/DiscussionPage.tsx
Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/DiscussionPage.tsx]: I blame the brutal merge conflict with too many changes 😤   but good catch 😄

**Comment by darrenvong** [comment]
> Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?

Wouldn't that break the pagination logic though, as the limit bit might be dependent on it? I think I'll leave this to the backend folks for that one... but yeah happy to treat it as done for the sake of this PR

>  it opens the comment box underneath all the other comments. We could do that!

I'll have a play around with that to see how much more complex that might make it than just placing it near the bottom and probably go with whatever's simpler. The extra complexity being needing to keep refs around probably to scroll to the comment box, but it might not be too bad

File: app/frontend/src/features/communities/discussions/CommentForm.tsx
Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/CommentForm.tsx:62]: We went with showing a refetch loading state instead, since optimistically updating the data in an infinite query is much more of a pain than a normal query. There is one in `Comment`, but I may have to adjust the position of it to be at the bottom if we are going with having the comment input near the last comment.

Comment (lucaslcode) [reply app/frontend/src/features/communities/discussions/CommentForm.tsx:62]: I see it now in storybook. For some reason I didn't before, maybe I just missed it.

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/Comment.tsx]: That's a nice trick actually rather than needing media queries. Still had to do an annoying workaround with '&' to select the same thing again and put in the clamp since you can't have duplicate object keys :(

**Comment by aapeliv** [comment]
> > Good point. Let's reverse it in the backend then - changing `.desc()` to `.asc()` in `threads.py` lines 67 and 90 should be enough, but it might break some tests, so leave it to another PR if you want. But treat it as if it has been done?
> 
> Wouldn't that break the pagination logic though, as the limit bit might be dependent on it? I think I'll leave this to the backend folks for that one... but yeah happy to treat it as done for the sake of this PR

See #1223

**Comment by darrenvong** [comment]
I think I addressed all the comments now apart from needing to add the "load earlier comments" button somewhere. Feel free to have another look through in between

**Comment by darrenvong** [comment]
Ok, list of messages is now reversed in the frontend, and the load more button is done now.

If you manage to check before I have more time to add Storybook (need to hack the mock a bit to deal with pagination), you'll probably have to pull this down locally and change the `pageSize` inside getThread to something smaller temporarily to see the button, but otherwise the functionality is pretty much there now

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (darrenvong) [inline app/frontend/src/features/communities/discussions/Comment.tsx]: Safari doesn't support smooth scrolling AND scrolling element to centre of the screen even with the polyfill, but calling `scrollIntoViewIfNeeded(true)` does it but without smooth scrolling. I thought for Safari only, it's worth giving up smooth scrolling for this, otherwise the input always scroll to top (default option) of screen and so the input gets clipped under the nav bar, which is probably more annoying than not having smooth scrolling imo.

Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/Comment.tsx]: Also since I'm not measuring anything and probably because of the 1-2s delay caused by the animation revealing the input, I managed to get away with using `useEffect` to do this rather than with `useLayoutEffect` :D

Comment (lucaslcode) [reply app/frontend/src/features/communities/discussions/Comment.tsx]: Maybe we need a better polyfill - does this one work? https://github.com/magic-akari/seamless-scroll-polyfill

Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/Comment.tsx]: Haven't tried yet but from looking at the code, it looks more promising and comprehensive than the original polyfill we are using now (which this is a fork of!)

Comment (darrenvong) [reply app/frontend/src/features/communities/discussions/Comment.tsx]: That other polyfill worked like a charm, nice find!

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/Comment.tsx]: Aha! I wasn't going crazy - this doesn't appear for the first nested comment, because it is inside `hasAtLeastOnePage`

File: app/frontend/package.json
Comment (lucaslcode) [inline app/frontend/package.json:26]: also remove the types dependency

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/Comment.tsx]: I was imagining this would also be a "Load earlier replies" button, but maybe it's not so important

**Review by lucaslcode — CHANGES_REQUESTED** [review]
It's great, just a few very minor things

File: app/frontend/src/features/communities/discussions/Comment.tsx
Comment (lucaslcode) [inline app/frontend/src/features/communities/discussions/Comment.tsx:147]: Can this also be moved to above the list of replies?

**Review by lucaslcode — CHANGES_REQUESTED** [review]
Last thing

**Review by lucaslcode — APPROVED** [review]
lgtm pending checks!

**Pull request merged**
